### PR TITLE
Ignore assembly version when activating DataProtection types from string name

### PR DESCRIPTION
**Pull request by natemcmaster**
Resolves https://github.com/aspnet/DataProtection/issues/187 by ignoring assembly version in the code the rehydrates types from strings.

Incidental fixups:

 - register activator in DI with TryAddSingleton
 - update Redis package to 0.3.0. It was building as 0.1.0, even though 0.1.1 has been released

### CODE REVIEW

File: src/Microsoft.AspNetCore.DataProtection/SimpleActivator.cs [hunk 1]
```diff
@@ -27,6 +27,8 @@ public SimpleActivator(IServiceProvider services)
 
         public virtual object CreateInstance(Type expectedBaseType, string implementationTypeName)
         {
+            implementationTypeName = RemoveVersionFromAssemblyName(implementationTypeName);
```
Comment (pakrym) [inline src/Microsoft.AspNetCore.DataProtection/SimpleActivator.cs]: Lets disable this on core.

File: src/Microsoft.AspNetCore.DataProtection/SimpleActivator.cs [hunk 2]
```diff
@@ -52,5 +54,33 @@ public virtual object CreateInstance(Type expectedBaseType, string implementatio
             // This will throw if the ctor cannot be called.
             return Activator.CreateInstance(implementationType);
         }
+
+        private static string RemoveVersionFromAssemblyName(string typeName)
+        {
+            var idx = typeName.IndexOf(',');
```
Comment (pakrym) [inline src/Microsoft.AspNetCore.DataProtection/SimpleActivator.cs]: Would this fail on generic types?

File: src/Microsoft.AspNetCore.DataProtection/SimpleActivator.cs [hunk 3]
```diff
@@ -52,5 +54,33 @@ public virtual object CreateInstance(Type expectedBaseType, string implementatio
...
+            if (idx <= 0 || (idx + 1) >= typeName.Length)
+            {
+                return typeName;
+            }
+
+            // type name includes assembly name
+            var assemblyName = typeName.Substring(idx + 1);
```
Comment (pakrym) [inline src/Microsoft.AspNetCore.DataProtection/SimpleActivator.cs]: This needs to be enabled only for predefined list of our assemblies.

File: test/Microsoft.AspNetCore.DataProtection.Test/ActivatorTests.cs [hunk 4]
```diff
@@ -2,14 +2,15 @@
...
+using System.Reflection;
 using Microsoft.AspNetCore.DataProtection.Internal;
 using Microsoft.Extensions.DependencyInjection;
 using Moq;
 using Xunit;
 
 namespace Microsoft.AspNetCore.DataProtection
 {
-    public class ActivatorTests
+    public class ActivatorTests : MarshalByRefObject
```
Comment (pakrym) [inline test/Microsoft.AspNetCore.DataProtection.Test/ActivatorTests.cs]: Why?

File: src/Microsoft.AspNetCore.DataProtection/SimpleActivator.cs (hunk 3 above)
Comment (natemcmaster) [inline src/Microsoft.AspNetCore.DataProtection/SimpleActivator.cs]: Should it be a specific list or is `AssemblyName.StartsWith("Microsoft.AspNetCore")` ok?

File: test/Microsoft.AspNetCore.DataProtection.Test/ActivatorTests.cs (hunk 4 above)
Comment (natemcmaster) [inline test/Microsoft.AspNetCore.DataProtection.Test/ActivatorTests.cs]: Required to make this work: AppDomain.CreateInstanceAndUnwrap.

File: src/Microsoft.AspNetCore.DataProtection/SimpleActivator.cs (hunk 2 above)
Comment (natemcmaster) [inline src/Microsoft.AspNetCore.DataProtection/SimpleActivator.cs]: I don't think so. I'll add a test to make sure

File: src/Microsoft.AspNetCore.DataProtection/SimpleActivator.cs (hunk 3 above)
Comment (pakrym) [inline src/Microsoft.AspNetCore.DataProtection/SimpleActivator.cs]: Let's do just DataProtection assemblies.

**Comment by natemcmaster** [comment]
🆙 :date:

**Comment by natemcmaster** [comment]
Btw, I verified the end-to-end scenario manually using this project file and the code provided in #187

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net461</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <!--<PackageReference Include="Microsoft.AspNetCore.DataProtection.Extensions" Version="2.0.0-*"/>-->
    <ProjectReference Include="../../src/Microsoft.AspNetCore.DataProtection.Extensions/Microsoft.AspNetCore.DataProtection.Extensions.csproj"/>
  </ItemGroup>

</Project>
```

**Comment by PinpointTownes** [comment]
@natemcmaster thanks for fixing this bug! :clap:

**Comment by pakrym** [comment]
Add a test to verify that version is not striped from non asp.net core assemblies.

**Comment by natemcmaster** [comment]
:up: :date:

**Pull request merged**